### ISSUE TITLE
Allow time varying delays in Synapses as an option

### ISSUE THREAD

**Issue by thesamovar** [brian-team/brian2#355]

**Comment by phreeza** [comment]
Are there any plans on integrating this feature? I think brian 1 had it, right? If so, would it be hard porting it from there? I would be interested in doing it.

**Comment by mstimberg** [comment]
Yes, we are certainly interested in this (just ignore that this issue was already opened 2 1/2 years ago ;) ). The feature was in Brian 1 as an experimental feature, but I don't think we necessarily have to port it from there. The basic change is relatively obvious, the difficulty comes from getting the details right.

The core reason why delays are not allowed to change during the run is that at the start of a run, we transform all delays into "offsets", i.e. simply integers representing the delay in timesteps (which is used to insert arriving spikes into the `SpikeQueue`): https://github.com/brian-team/brian2/blob/master/brian2/synapses/cspikequeue.cpp#L90

The easiest way to implement support for changing delays would be to make this preparation optional and only do it when the user uses constant delays, otherwise we transform the delay into an integer value for each spike. It would probably a good idea to profile this a bit and see whether it makes a meaningful speed difference (by comparing two simulations where delays are constant, but in one case, they are marked as variable). If it does not, we could also *never* pre-calculate integer delays. On the other hand, if it makes a big difference, we might e.g. keep a look-up table to replace a relatively  costly division/rounding operation by a look-up.

Then, there are a few syntax questions: should it be an explicit choice by the user to allow the use of variable delays, or do we switch automatically if there is any code that modifies the delays? I'd probably vote for the former, but then again if we decide that pre-calculation does not save us anything, we could simply always allow variable delays. Finally, how much do we check the delay values that the user dynamically sets in their code? Setting negative values would certainly give weird results, but our code-generation framework has no general mechanism to control for this [1]. Also, our `SpikeQueue` data structure is based on a maximum delay -- what happens if the dynamic change makes a delay bigger than the previously maximal delay? Do we have a dynamic data structure with a bit of buffer for this situation (e.g. like a dynamic array, whenever we exceed the current maximal delay, we double it)? Or does the user need to manually set a maximal delay in the beginning? The former can be potentially costly in memory, and the latter can be possibly annoying if it raises an error for delays exceeding the set maximum value (imagine this error at some point during a long-running simulation just because a single delay became a tiny bit bigger than the maximal delay), or lead to hard-to-find bugs if it silently clips delays to the maximum.

One thing that we try to do when adding new features like this is to first find one or more specific use cases (e.g. do you know of a model that uses variable delays, e.g. a model where delays evolve according to a learning rule?) -- this helps a lot when deciding about the feature scope and the syntax.

Hopefully I did not discourage you to work on this by this long "there are a lot of things to consider" comment :worried: It would be certainly great if you could work on this, but let's make sure we have found a consensus on the general questions that I outlined above, first.

[1] In general, we try to avoid adding hard-coded support for a very specific feature like this in the code code generation framework. However, we could first implement a general support for variable value limits (#377) and then delays would be just one use case.

**Comment by phreeza** [comment]
Thanks for this detailed reply :) I agree it is a good idea to get these details fleshed out first. I will weigh in where I can:

> Do we have a dynamic data structure with a bit of buffer for this situation (e.g. like a dynamic array, whenever we exceed the current maximal delay, we double it)?

I think a dynamic array that doubles as needed is the best way to go here. The space overhead is not really that much, it is at most a factor of 2 worse than the static version, right? And allocation has O(1) amortised time complexity. Another option would be a sparse data structure instead, ie an actual queue of spikes and the time they were emitted. 

Another question that I am unclear on: How is a spike handled if the delay is changed after the spike is generated? Is there a single SpikeQueue for every postsynaptic neuron (in which case a spike would need to keep the delay it had when it was fired) or is there a SpikeQueue for every single synapse? If spike delays can change after firing, there would also be a number of edge cases to consider: What happens if delay changes such that a spike is "skipped"? Even weirder: what happens if a spike arrives but then the delay changes such that the delay is again such that the spike arrives in the future. Is it fired again? Or does it disappear?

I also saw there is #467, which might be an even more elegant way of solving this, if a synapse with a delay is just an instantaneous synapse plus access to a delayed version of the presynaptic variable?

> Setting negative values would certainly give weird results, but our code-generation framework has no general mechanism to control for this [1]. 

I think having a way to raise errors from within generated code would definitely be worthwhile, probably also for other parts of the code? But apart from that, I think clipping negative delays to zero should be fine, and pretty intuitive. Perhaps emit a warning?

> One thing that we try to do when adding new features like this is to first find one or more specific use cases (e.g. do you know of a model that uses variable delays, e.g. a model where delays evolve according to a learning rule?) -- this helps a lot when deciding about the feature scope and the syntax.

What I am looking at implementing is something along the lines of [this publication](https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.82.1594)

**Comment by mstimberg** [comment]
> I think a dynamic array that doubles as needed is the best way to go here. The space overhead is not really that much, it is at most a factor of 2 worse than the static version, right? And allocation has O(1) amortised time complexity. Another option would be a sparse data structure instead, ie an actual queue of spikes and the time they were emitted.

Right, I agree a dynamic structure would be best. I kind of had in mind that the `SpikeQueue` has a size on the order of the number of synapses, but this is not at all the case. And in fact the data structure we use is already dynamic (it's a STL vector of length "possible integer delays", with each referring an STL vector of the synapses that spiked). The only thing that is missing is to allow for an increase in queue size during a `push` (and not only during the initial preparation), and moving things accordingly (it is a circular data structure). Sorry about this confusion.

> Another question that I am unclear on: How is a spike handled if the delay is changed after the spike is generated?

Yes, this is an important question. There are basically two approaches to spike queues: one where you store all spikes for a current time step in the same way, but check for previously happened spikes according to the delays. The second approach is to store spikes of the current time step according to the respective delays (i.e. in time slots representing future time steps), and looking up all spikes that are supposed to be delivered at the current time step. Brian uses the latter convention, therefore there is a straightforward answer to the question: the delay matters at the time of the spike emission, latter delay changes do no affect spikes that are already in the pipeline. This nicely avoids weird effects such as skipped spikes, etc.

> I also saw there is #467, which might be an even more elegant way of solving this, if a synapse with a delay is just an instantaneous synapse plus access to a delayed version of the presynaptic variable?

No, it isn't, but that's actually a bigger issue that we have to address at some point. You have pre-synaptic and post-synaptic delays and they determine what is considered to be "synchronous" from the synapse point of view. For example, if the pre-synaptic delay is 3ms and the post-synaptic delay is 5ms, then if the pre- and post-synaptic cells spike synchronously, this will be considered as a "post-after-pre" pairing at the synapse, which is obviously quite important for STDP models. However, the post-synaptic delay is only a kind of "backpropagation delay". In contrast to some other simulators (NEST, for example, IIRC), for the propagation to the post-synaptic cell only the pre-synaptic delay is considered, making it an "axonal delay" and not a "dendritic delay": after a certain delay, the `on_pre` action is calculated and applied immediately to the post-synaptic cell. Where this becomes less intuitive is when referring to pre-synaptic variables: they are also evaluated at the time of the `on_pre` evaluation -- other simulators send the value of the pre-synaptic variable at the spike time as a "payload" information with the spike, which -- depending on what your pre-synaptic variable represents -- might be more intuitive or not. There are ways around this under certain conditions (e.g. if the refractory time of the pre-synaptic cell is bigger than the delay then you can use two pathways to simulate dendritic delays), but it's not straightforward in general. Well, I digressed quite a bit...

> I think having a way to raise errors from within generated code would definitely be worthwhile, probably also for other parts of the code? But apart from that, I think clipping negative delays to zero should be fine, and pretty intuitive. Perhaps emit a warning?

Many possible options, I'd be fine with clipping/warnings/etc., but I'd prefer it to be a general mechanism like the one proposed in #377. At the moment, at the code generation level, `delay` is a synaptic variable as any other variable (e.g. the weight), and I'd like to keep it that way, i.e. not introduce anything that explicitly checks whether the variable that the user sets in their, say, `on_pre` statement is a delay variable and we should therefore clip it.

> What I am looking at implementing is something along the lines of this publication

Thanks, that should be helpful. From quickly skimming it, it basically uses an STDP rule, but instead of changing the weight, it changes the delay, right?

**Comment by thesamovar** [comment]
I still think this would be a really nice feature to have. I was sad to lose this from Brian 1 (but it was a sensible decision in terms of priorities).

Pretty much agree with @mstimberg, just a couple of extra points.

For the issue of whether or not the delay of a spike changes, in addition to what you have both said that makes it ill defined and potentially acausal if you try to change the delay while the spike is en route, another point is that it probably shouldn't change so fast that it matters. Changes in delay would presumably be associated with rather slow growth processes, so very rapid changes won't happen.

Seems to me like generating errors/warnings from within generated code seems sensible. Do we already have this? If not, maybe worth adding an issue on that.

**Comment by phreeza** [comment]
Thanks for the clarifications. I have started reading through the code to really grok how the Synapse objects work, I had not realized there are separate objects for the pre and post pathway.

> From quickly skimming it, it basically uses an STDP rule, but instead of changing the weight, it changes the delay, right?
Yes, that is how i think about it as well.

> For the issue of whether or not the delay of a spike changes, in addition to what you have both said that makes it ill defined and potentially acausal if you try to change the delay while the spike is en route, another point is that it probably shouldn't change so fast that it matters. Changes in delay would presumably be associated with rather slow growth processes, so very rapid changes won't happen.

Slowness seems like a reasonable assumption, but I think one could still get such effects due to the discretization of the time steps, eg if dt = 1ms, and a spike at t=2ms delay shifts from 2.99ms at t=4ms to 3.01ms at t=5ms, it could still be counted twice.

Do you have any concrete suggestions what I should do to get started? Should I care about #377/variable limits or should that stay a separate issue?

**Comment by mstimberg** [comment]
> Slowness seems like a reasonable assumption, but I think one could still get such effects due to the discretization of the time steps, eg if dt = 1ms, and a spike at t=2ms delay shifts from 2.99ms at t=4ms to 3.01ms at t=5ms, it could still be counted twice.

Right, but as I mentioned earlier, I think all of this should not matter if we use the straightforward approach: the delay applies at the time when the spike is *emitted*, later delay changes do not affect spikes that are already "on their way".

> Do you have any concrete suggestions what I should do to get started? Should I care about #377/variable limits or should that stay a separate issue?

I think I'd tackle the two issues separately, even though I would prefer to only do release a Brian version with support for dynamic delays if issue #377 (or some other mechanism to avoid negative delays) has been addressed as well. If you are interested in working on both, then it's up to you which one to work on first. Also, we could discuss the syntax/semantics for #377 in the issue comments while you work on this issue here in parallel?
